lib/srv/forward/sshserver.go

```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+	sch, sin, err := xreq.Accept()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// setup outbound X11 channel to client
+	cch, cin, err := s.sconn.OpenChannel(sshutils.X11ChannelRequest, xreq.ExtraData())
+	if err != nil {
+		sch.Close()
+		return trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// scw and ccw protect CloseWrite calls against sch and cch respectively.
+	var scw, ccw sync.Once
+
+	// ioflag is used by the two io goroutines to determine if they are the
+	// first or second goroutine to complete.  The second goroutine to complete
+	// ensures that the main request forwarding goroutine also terminates by
+	// cancelling its context.
+	var ioflag uint32
+
+	defer func() {
+		// consume our sync.Once instances if they haven't been
+		// consumed already (prevents CloseWrite calls from racing
+		// with Close calls).
+		scw.Do(func() {})
+		ccw.Do(func() {})
+		cancel()
+		sch.Close()
+		cch.Close()
+	}()
+
+	go func() {
+		// forward data from server to client
+		io.Copy(cch, sch)
+		// inform client that no more data is coming
+		ccw.Do(func() {
+			cch.CloseWrite()
+		})
+		// if we are the second io forwarder to complete,
+		// then signal shutdown to main goroutine.
+		if atomic.SwapUint32(&ioflag, 1) > 0 {
+			cancel()
+		}
+	}()
+
+	go func() {
+		// forward data from client to server
+		io.Copy(sch, cch)
+		// inform server that no more data is coming
+		scw.Do(func() {
+			sch.CloseWrite()
+		})
+		// if we are the second io forwarder to complete,
+		// then signal shutdown to main goroutine.
+		if atomic.SwapUint32(&ioflag, 1) > 0 {
+			cancel()
+		}
+	}()
+
+	// multiplex requests until channel is closed or both
+	// io forwarders have completed.
+	for {
+		select {
+		case sreq := <-sin:
+			if sreq == nil {
+				return nil
+			}
+			switch sreq.Type {
+			case sshutils.WindowChangeRequest:
+				if _, err := forwardRequest(cch, sreq); err != nil {
+					return trace.Wrap(err)
+				}
+			default:
+				s.log.Errorf("Unsupported X11 channel request type: %s", sreq.Type)
+				if sreq.WantReply {
+					sreq.Reply(false, nil)
+				}
+				continue
+			}
+		case creq := <-cin:
+			if creq == nil {
+				return nil
+			}
+			switch creq.Type {
+			case sshutils.WindowChangeRequest:
+				if _, err := forwardRequest(sch, creq); err != nil {
+					return trace.Wrap(err)
+				}
+			default:
+				s.log.Errorf("Unsupported X11 channel request type: %s", creq.Type)
+				if creq.WantReply {
+					creq.Reply(false, nil)
+				}
+				continue
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// serveX11Channels registers and runs a hander for X11 channel requests
+// generated by the server.  Must not be launched until X11 forwarding has
+// been requested and authorized.
+func (s *Server) serveX11Channels(ctx context.Context) error {
+	ncs := s.remoteClient.HandleChannelOpen(sshutils.X11ChannelRequest)
+	if ncs == nil {
+		return trace.AlreadyExists("x11 forwarding already active")
```

Comment:
nit: `s/x11/X11` to be consistent

Reply:
I'm deliberately using `x11` instead of `X11` in places where proper nouns are typically lowercase (e.g. error messages, config fields).  This is consistent with how openssh handles it, and with how teleport generally treats terms like `http`/`HTTP`.